### PR TITLE
Fix crashes in config parsing

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by bluca** [comment]
> Assertion 'format' failed at src/core/unit-printf.c:232, function unit_full_printf(). Aborting.

File: src/core/load-fragment.c [hunk 1]
```diff
@@ -5684,21 +5684,27 @@ int config_parse_cgroup_socket_bind(
         r = extract_first_word(&rvalue, &word, ":", 0);
         if (r == -ENOMEM)
                 return log_oom();
+        if (r <= 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Unable to parse %s assignment, ignoring: %s", lvalue, rvalue);
```
Comment (yuwata) [inline src/core/load-fragment.c]: Maybe `%s=` (i.e., missing '=')?

File: src/core/load-fragment.c [hunk 2]
```diff
@@ -5684,21 +5684,27 @@ int config_parse_cgroup_socket_bind(
...
-        if (address_family) {
-                if (streq(address_family, "IPv4"))
+        if (rvalue) {
+                if (streq(word, "IPv4"))
                         af = AF_INET;
-                else if (streq(address_family, "IPv6"))
+                else if (streq(word, "IPv6"))
                         af = AF_INET6;
                 else
                         return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
```
Comment (yuwata) [inline src/core/load-fragment.c]: Please use `log_syntax()`, and it is better to return `0`.
The same pattern appears after `parse_ip_port_range()` below. Could you also update it?

File: src/core/load-fragment.c [hunk 3]
```diff
@@ -5630,7 +5630,7 @@ int config_parse_bpf_foreign_program(
         r = extract_first_word(&rvalue, &word, ":", 0);
         if (r == -ENOMEM)
                 return log_oom();
-        if (r < 0) {
+        if (r <= 0) {
```
Comment (yuwata) [inline src/core/load-fragment.c]: Could you also update `log_error_errno()` -> `log_syntax()` and return `0` after `cgroup_add_bpf_foreign_program() below.`

**Review by yuwata — CHANGES_REQUESTED** [review]
Could you update the commit message of the last commit?
`core: fix crash in BPFProgram parsing` -> `fuzz: add test case for BPFProgram= parsing` or something.

**Comment by keszybz** [comment]
> > Assertion 'format' failed at src/core/unit-printf.c:232, function unit_full_printf(). Aborting.

Yeah, I managed to push the reproducer without the fix, smart ;]

File: src/core/load-fragment.c (hunk 3 above)
Comment (keszybz) [reply src/core/load-fragment.c]: Updated, except that `cgroup_add_bpf_foreign_program()` can only fails with -ENOMEM, so I left it as it was.

**Comment by bluca** [comment]
> > > Assertion 'format' failed at src/core/unit-printf.c:232, function unit_full_printf(). Aborting.
> 
> Yeah, I managed to push the reproducer without the fix, smart ;]

At least we know the reproducer works :-)

**Comment by keszybz** [comment]
Should be OK now. I made the requested changes, except as commented above.

File: src/core/load-fragment.c (hunk 3 above)
Comment (yuwata) [reply src/core/load-fragment.c]: Ah, I see. Thanks.

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/libsystemd/sd-device/sd-device.c [hunk 4]
```diff
@@ -245,29 +245,31 @@ _public_ int sd_device_new_from_devnum(sd_device **ret, char type, dev_t devnum)
...
 
         assert_return(ret, -EINVAL);
         assert_return(subsystem, -EINVAL);
         assert_return(sysname, -EINVAL);
+        assert_return(strlen(sysname) < PATH_MAX - strlen("/sys/bus/"), -ENAMETOOLONG);
 
         if (streq(subsystem, "subsystem")) {
-                syspath = strjoina("/sys/subsystem/", sysname);
-                if (access(syspath, F_OK) >= 0)
+                if (snprintf_ok(syspath, sizeof syspath, "/sys/subsystem/%s", sysname) &&
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:256]: snprintf_ok() is a lot more expensive thatn strjoina().

why not instead do a filename_is_valid() on the two args first and then use strjoina() as before? the filename_is_valid() does a length check after all, and the stuff should be validated then, which i a good idea anyway

**Pull request merged**